### ISSUE TITLE
Numeric Value out of range when sending floats and numeric types to Sybase ASE 16 #223

### ISSUE THREAD

**Issue by qlecler** [FreeTDS/freetds#285]
Hi there!

I've a similar issue as [#223](https://github.com/FreeTDS/freetds/issues/223)

Is it fixed yet?

Thanks

**Comment by freddy77** [comment]
I think the reported closed it.
What do you mean similar?
Can you post some code?

**Comment by qlecler** [comment]
Yes with a sample code like this : 

`  UPDATE credit SET mnt_prime = 33.44`

I get this error :

```
10:38:42.874642 36633 (error.c:414):odbc_errs_add: "Numeric value out of range"
10:38:42.874652 36633 (odbc.c:3590):SQLExecute returns SQL_ERROR (start_parse_prepared_query failed)
10:38:42.874662 36633 (error_export.h:19):SQLGetDiagRecW(3, 0x10320abf0, 1, 0x70000c9584d4, 0x70000c9584cc, 0x70000c9584e0, 512, 0x70000c9584d0)
10:38:42.874674 36633 (error.c:559):SQLGetDiagRec: "[FreeTDS][SQL Server]Numeric value out of range"
```

Theses queries worked without issues :

`  UPDATE credit SET mnt_prime = 33`
`  UPDATE credit SET mnt_prime = 33.00`

**Comment by freddy77** [comment]
> Yes with a query like this :
> 
> ` UPDATE credit SET mnt_prime = 33.44`
> 

I suppose this is not a simple string you pass to SQLExecDirect. How do you bind and pass the `33.44` ?

**Comment by qlecler** [comment]
> > Yes with a query like this :
> > ` UPDATE credit SET mnt_prime = 33.44`
> 
> I suppose this is not a simple string you pass to SQLExecDirect. How do you bind and pass the `33.44` ?

I'm using ODBC (for Node.js) and I've try to execute this query directly, It doesn't work either.

**Comment by freddy77** [comment]
You should enable ODBC trace logs and send the part that send this query.

**Comment by qlecler** [comment]
```
16:31:29.073906 41983 (odbc_export.h:837):SQLPrepareW(0x102709a40,
  UPDATE credit_cours SET id_unique = ?, type_credit = ?, Rembourse_credit = ?, taeg = ?, solde = ?, mnt_Credit = ?, mnt_echeance = ?, indemnite_reemploi = ?, duree_credit = ?, date_premiere_echeance = ?,  frais_mainlevee = ? WHERE id_credit_cours = ?
, 253)
16:31:29.073938 41983 (odbc.c:4594):Creating prepared statement
16:31:29.073950 41983 (mem.c:656):tds_free_all_results()
16:31:29.073961 41983 (util.c:165):Changed query state from IDLE to WRITING
16:31:29.073988 41983 (packet.c:801):Sending packet
0000 0f 01 01 17 00 00 00 00-e7 0c 01 01 00 0a 61 38 |........ ......a8|
0010 6b 31 64 72 38 30 30 30-fd 00 0a 20 20 55 50 44 |k1dr8000 ...  UPD|
0020 41 54 45 20 63 72 65 64-69 74 5f 63 6f 75 72 73 |ATE cred it_cours|
0030 20 53 45 54 20 69 64 5f-75 6e 69 71 75 65 20 3d | SET id_ unique =|
0040 20 3f 2c 20 74 79 70 65-5f 63 72 65 64 69 74 20 | ?, type _credit |
0050 3d 20 3f 2c 20 52 65 6d-62 6f 75 72 73 65 5f 63 |= ?, Rem bourse_c|
0060 72 65 64 69 74 20 3d 20-3f 2c 20 74 61 65 67 20 |redit =  ?, taeg |
0070 3d 20 3f 2c 20 73 6f 6c-64 65 20 3d 20 3f 2c 20 |= ?, sol de = ?, |
0080 6d 6e 74 5f 43 72 65 64-69 74 20 3d 20 3f 2c 20 |mnt_Cred it = ?, |
0090 6d 6e 74 5f 65 63 68 65-61 6e 63 65 20 3d 20 3f |mnt_eche ance = ?|
00a0 2c 20 69 6e 64 65 6d 6e-69 74 65 5f 72 65 65 6d |, indemn ite_reem|
00b0 70 6c 6f 69 20 3d 20 3f-2c 20 64 75 72 65 65 5f |ploi = ? , duree_|
00c0 63 72 65 64 69 74 20 3d-20 3f 2c 20 64 61 74 65 |credit =  ?, date|
00d0 5f 70 72 65 6d 69 65 72-65 5f 65 63 68 65 61 6e |_premier e_echean|
00e0 63 65 20 3d 20 3f 2c 20-20 66 72 61 69 73 5f 6d |ce = ?,   frais_m|
00f0 61 69 6e 6c 65 76 65 65-20 3d 20 3f 20 57 48 45 |ainlevee  = ? WHE|
0100 52 45 20 69 64 5f 63 72-65 64 69 74 5f 63 6f 75 |RE id_cr edit_cou|
0110 72 73 20 3d 20 3f 0a   -                        |rs = ?.|

16:31:29.074123 41983 (util.c:165):Changed query state from WRITING to PENDING
16:31:29.074137 41983 (token.c:553):tds_process_tokens(0x102a23fe0, 0x700008405cc8, 0x700008405ccc, 0x104)
16:31:29.074147 41983 (util.c:165):Changed query state from PENDING to READING
16:31:29.075308 41983 (packet.c:435):Received packet
0000 04 01 00 92 00 00 00 00-e7 0d 00 20 00 0a 61 38 |........ ... ..a8|
0010 6b 31 64 72 38 30 30 30-ec 6e 00 0c 00 01 3f 00 |k1dr8000 .n....?.|
0020 3a 00 00 00 38 00 01 3f-00 3a 00 00 00 38 00 01 |:...8..? .:...8..|
0030 3f 00 3a 00 00 00 38 00-01 3f 00 3a 00 00 00 38 |?.:...8. .?.:...8|
0040 00 01 3f 00 3a 00 00 00-38 00 01 3f 00 3a 00 00 |..?.:... 8..?.:..|
0050 00 38 00 01 3f 00 3a 00-00 00 38 00 01 3f 00 3a |.8..?.:. ..8..?.:|
0060 00 00 00 38 00 01 3f 00-3a 00 00 00 38 00 01 3f |...8..?. :...8..?|
0070 00 3a 00 00 00 38 00 01-3f 00 3a 00 00 00 38 00 |.:...8.. ?.:...8.|
0080 01 3f 00 3a 00 00 00 38-00 fd 00 00 01 00 00 00 |.?.:...8 ........|
0090 00 00                  -                        |..|

16:31:29.075390 41983 (token.c:568):processing result tokens.  marker is  e7(TDS5_DYNAMIC)
16:31:29.075412 41983 (token.c:568):processing result tokens.  marker is  ec(TDS5_PARAMFMT)
16:31:29.075427 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21590, 1) [for parameter]
16:31:29.075439 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075449 41983 (token.c:1712):processing result. column_size 4
16:31:29.075458 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21650, 1) [for parameter]
16:31:29.075468 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075477 41983 (token.c:1712):processing result. column_size 4
16:31:29.075485 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21710, 1) [for parameter]
16:31:29.075495 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075503 41983 (token.c:1712):processing result. column_size 4
16:31:29.075519 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c217d0, 1) [for parameter]
16:31:29.075532 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075542 41983 (token.c:1712):processing result. column_size 4
16:31:29.075568 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21890, 1) [for parameter]
16:31:29.075592 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075605 41983 (token.c:1712):processing result. column_size 4
16:31:29.075613 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21950, 1) [for parameter]
16:31:29.075623 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075641 41983 (token.c:1712):processing result. column_size 4
16:31:29.075655 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21a10, 1) [for parameter]
16:31:29.075665 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075691 41983 (token.c:1712):processing result. column_size 4
16:31:29.075700 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21ad0, 1) [for parameter]
16:31:29.075710 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075738 41983 (token.c:1712):processing result. column_size 4
16:31:29.075746 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21bb0, 1) [for parameter]
16:31:29.075755 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075763 41983 (token.c:1712):processing result. column_size 4
16:31:29.075771 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21c70, 1) [for parameter]
16:31:29.075780 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075789 41983 (token.c:1712):processing result. column_size 4
16:31:29.075798 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21d30, 1) [for parameter]
16:31:29.075807 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075815 41983 (token.c:1712):processing result. column_size 4
16:31:29.075824 41983 (token.c:1645):tds_get_data_info(0x102a23fe0, 0x102c21df0, 1) [for parameter]
16:31:29.075833 41983 (token.c:1707):processing result. type = 56(int), varint_size 0
16:31:29.075842 41983 (token.c:1712):processing result. column_size 4
16:31:29.075852 41983 (token.c:568):processing result tokens.  marker is  fd(DONE)
16:31:29.075862 41983 (token.c:2094):tds_process_end: more_results = 0
		was_cancelled = 0
		error = 0
		done_count_valid = 0
16:31:29.075871 41983 (token.c:2113):                rows_affected = 0
16:31:29.075880 41983 (token.c:2116):tds_process_end() state set to TDS_IDLE
16:31:29.075889 41983 (util.c:165):Changed query state from READING to IDLE
16:31:29.075899 41983 (util.c:83):logic error: cannot change query state from IDLE to PENDING
16:31:29.075908 41983 (util.c:165):Changed query state from IDLE to IDLE
16:31:29.075917 41983 (token.c:553):tds_process_tokens(0x102a23fe0, 0x700008405cc8, 0x700008405ccc, 0x104)
16:31:29.075926 41983 (token.c:556):tds_process_tokens() state is COMPLETED
16:31:29.076368 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 1, 1, -25, -5, 0, 0, 0x102707c30, 0, 0x10270a288)
16:31:29.076393 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 1, 1, -25, -5, 0, 0, 0x102707c30, 0, 0x10270a288)
16:31:29.076410 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 2, 1, -8, -9, 0, 0, 0x1027021d0, 10, 0x10270a2b8)
16:31:29.076420 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 2, 1, -8, -9, 0, 0, 0x1027021d0, 10, 0x10270a2b8)
16:31:29.076432 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 3, 1, -8, -9, 0, 0, 0x1027021e0, 4, 0x10270a2e8)
16:31:29.076442 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 3, 1, -8, -9, 0, 0, 0x1027021e0, 4, 0x10270a2e8)
16:31:29.076458 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 4, 1, -8, -9, 0, 0, 0x1027021f0, 4, 0x10270a318)
16:31:29.076469 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 4, 1, -8, -9, 0, 0, 0x1027021f0, 4, 0x10270a318)
16:31:29.076480 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 5, 1, -8, -9, 0, 0, 0x102707df0, 14, 0x10270a348)
16:31:29.076490 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 5, 1, -8, -9, 0, 0, 0x102707df0, 14, 0x10270a348)
16:31:29.076500 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 6, 1, -8, -9, 0, 0, 0x102707e00, 12, 0x10270a378)
16:31:29.076509 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 6, 1, -8, -9, 0, 0, 0x102707e00, 12, 0x10270a378)
16:31:29.076519 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 7, 1, -8, -9, 0, 0, 0x102707e10, 14, 0x10270a3a8)
16:31:29.076528 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 7, 1, -8, -9, 0, 0, 0x102707e10, 14, 0x10270a3a8)
16:31:29.076537 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 8, 1, -8, -9, 0, 0, 0x102707e20, 12, 0x10270a3d8)
16:31:29.076545 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 8, 1, -8, -9, 0, 0, 0x102707e20, 12, 0x10270a3d8)
16:31:29.076556 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 9, 1, -25, -5, 0, 0, 0x1027005e0, 0, 0x10270a408)
16:31:29.076583 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 9, 1, -25, -5, 0, 0, 0x1027005e0, 0, 0x10270a408)
16:31:29.076594 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 10, 1, -8, -9, 0, 0, 0x1027005f0, 22, 0x10270a438)
16:31:29.076604 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 10, 1, -8, -9, 0, 0, 0x1027005f0, 22, 0x10270a438)
16:31:29.076616 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 11, 1, 8, 3, 8, 7, 0x102700610, 8, 0x10270a468)
16:31:29.076626 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 11, 1, 8, 3, 8, 7, 0x102700610, 8, 0x10270a468)
16:31:29.076636 41983 (odbc.c:1509):SQLBindParameter(0x102709a40, 12, 1, -8, -9, 0, 0, 0x102709c50, 16, 0x10270a498)
16:31:29.076646 41983 (odbc.c:1410):_SQLBindParameter(0x102709a40, 12, 1, -8, -9, 0, 0, 0x102709c50, 16, 0x10270a498)
16:31:29.076807 41983 (odbc.c:3576):SQLExecute(0x102709a40)
16:31:29.076823 41983 (prepare_query.c:202):parsing 0 parameters
16:31:29.076835 41983 (sql2tds.c:184):type=-5
16:31:29.076845 41983 (sql2tds.c:192):trace
16:31:29.076866 41983 (sql2tds.c:184):type=-9
16:31:29.076876 41983 (sql2tds.c:192):trace
16:31:29.076904 41983 (sql2tds.c:184):type=-9
16:31:29.076918 41983 (sql2tds.c:192):trace
16:31:29.076929 41983 (sql2tds.c:184):type=-9
16:31:29.076938 41983 (sql2tds.c:192):trace
16:31:29.076949 41983 (sql2tds.c:184):type=-9
16:31:29.076958 41983 (sql2tds.c:192):trace
16:31:29.076968 41983 (sql2tds.c:184):type=-9
16:31:29.076977 41983 (sql2tds.c:192):trace
16:31:29.076988 41983 (sql2tds.c:184):type=-9
16:31:29.076998 41983 (sql2tds.c:192):trace
16:31:29.077008 41983 (sql2tds.c:184):type=-9
16:31:29.077017 41983 (sql2tds.c:192):trace
16:31:29.077025 41983 (sql2tds.c:184):type=-5
16:31:29.077031 41983 (sql2tds.c:192):trace
16:31:29.077038 41983 (sql2tds.c:184):type=-9
16:31:29.077044 41983 (sql2tds.c:192):trace
16:31:29.077051 41983 (sql2tds.c:184):type=3
16:31:29.077056 41983 (sql2tds.c:192):trace
16:31:29.077068 41983 (error.c:414):odbc_errs_add: "Numeric value out of range"
16:31:29.077074 41983 (odbc.c:3590):SQLExecute returns SQL_ERROR (start_parse_prepared_query failed)
16:31:29.077086 41983 (error_export.h:19):SQLGetDiagRecW(3, 0x102709a40, 1, 0x7000073ff4d4, 0x7000073ff4cc, 0x7000073ff4e0, 512, 0x7000073ff4d0)
16:31:29.077098 41983 (error.c:559):SQLGetDiagRec: "[FreeTDS][SQL Server]Numeric value out of range"
16:31:29.077106 41983 (error_export.h:19):SQLGetDiagRecW(3, 0x102709a40, 2, 0x7000073ff4d4, 0x7000073ff4cc, 0x7000073ff4e0, 512, 0x7000073ff4d0)
16:31:29.077152 41983 (error_export.h:19):SQLGetDiagRecW(3, 0x102709a40, 1, 0x7ffeefbf4512, 0x7ffeefbf3cd4, 0x7ffeefbf3d10, 1024, 0x7ffeefbf3d06)
16:31:29.077163 41983 (error.c:559):SQLGetDiagRec: "[FreeTDS][SQL Server]Numeric value out of range"
```

**Comment by freddy77** [comment]
I don't know why the reported closed #223.
However it's a bug of NodeJs odbc. Is using this code to set the type (probably if you don't have decimals it's using integers):
```cpp
    else if (value->IsNumber()) {
      double *number   = new double(value->NumberValue());

      params[i].ValueType         = SQL_C_DOUBLE;
      params[i].ParameterType     = SQL_DECIMAL;
      params[i].ParameterValuePtr = number;
      params[i].BufferLength      = sizeof(double);
      params[i].StrLen_or_IndPtr  = params[i].BufferLength;
      params[i].DecimalDigits     = 7;
      params[i].ColumnSize        = sizeof(double);

      DEBUG_PRINTF("ODBC::GetParametersFromArray - IsNumber(): params[%i] c_type=%i type=%i buffer_length=%lli size=%lli length=%lli value=%f\n",
                    i, params[i].ValueType, params[i].ParameterType,
                    params[i].BufferLength, params[i].ColumnSize, params[i].StrLen_or_IndPtr,
                                *number);
    }
```
Then is using as
```c
        ret = SQLBindParameter(
          hSTMT,                    //StatementHandle
          i + 1,                    //ParameterNumber
          SQL_PARAM_INPUT,          //InputOutputType
          prm.ValueType,
          prm.ParameterType,
          prm.ColumnSize,
          prm.DecimalDigits,
          prm.ParameterValuePtr,
          prm.BufferLength,
          &params[i].StrLen_or_IndPtr);
```
basically is passing 8 as precision and 7 as decimal that is a number like `X.XXXXXXX`. Correctly `33.44` is not accepted as `33` cannot fit in `X`. They probably want `SQL_REAL` instead of `SQL_DOUBLE`.
